src/elbo.jl

```diff
@@ -0,0 +1,244 @@
+end
+
+
+"""
+    elbo(lfx::LatentFiniteGP, y::AbstractVector, fz::FiniteGP, q::AbstractMvNormal; n_data=length(y), method=Default())
+
+Compute the ELBO for a LatentGP with a possibly non-conjugate likelihood.
+"""
+function elbo(
+    lfx::LatentFiniteGP,
+    y::AbstractVector,
+    fz::FiniteGP,
+    q::AbstractMvNormal;
+    n_data=length(y),
+    method=Default()
+)
+    return _elbo(method, lfx.fx, y, fz, q, lfx.lik, n_data)
+end
+
+# Compute the common elements of the ELBO
+function _elbo(
+    method::ExpectationMethod,
+    fx::FiniteGP,
+    y::AbstractVector,
+    fz::FiniteGP,
+    q::AbstractMvNormal,
+    lik::ScalarLikelihood,
+    n_data::Integer
+)
+    post = approx_posterior(SVGP(), fz, q)
+    f_mean, f_var = mean_and_var(post, fx.x)
+    variational_exp = expected_loglik(method, y, f_mean, f_var, lik)
+
+    kl_term = StatsBase.kldivergence(q, fz)
+
+    n_batch = length(y)
+    scale = n_data / n_batch
+    return sum(variational_exp) * scale - kl_term
+end
+
+"""
+    expected_loglik(method, y, f_mean, f_var, lik)
+
+This function computes the expected log likelihood:
+
+```math
+    ∫ q(f) log p(y | f) df
+```
+where `p(y | f)` is the process likelihood.
+
+`q(f)` is an approximation to the latent function values `f` given by:
+```math
+    q(f) = ∫ p(f | u) q(u) du
+```
+where `q(u)` is the variational distribution over inducing points (see
+[`elbo`](@ref)). The marginal means and variances of `q(f)` are given by
+`f_mean` and `f_var` respectively.
+
+`method` determines which method is used to calculate the expected log
+likelihood - see [`elbo`](@ref) for more details.
+
+# Extended help
+
+`q(f)` is assumed to be an `MvNormal` distribution and `p(y | f)` is assumed to
+have independent marginals such that only the marginals of `q(f)` are required.
+"""
+function expected_loglik end
+
+"""
+    expected_loglik(method::ExpectationMethod, y::AbstractVector, f_mean::AbstractVector, f_var::AbstractVector, lik::ScalarLikelihood)
+
+The expected log likelihood for a `ScalarLikelihood`, computed via `method`.
+Defaults to a closed form solution if it exists, otherwise defaults to
+Gauss-Hermite quadrature.
+"""
+function expected_loglik(
+    ::Default,
+    y::AbstractVector,
+    f_mean::AbstractVector,
+    f_var::AbstractVector,
+    lik::ScalarLikelihood
+)
+    method = _default_method(lik)
+    expected_loglik(method, y, f_mean, f_var, lik)
+end
+
+# The closed form solution for independent Gaussian noise
+function expected_loglik(
+    ::Analytic,
+    y::AbstractVector,
+    f_mean::AbstractVector,
+    f_var::AbstractVector,
+    lik::GaussianLikelihood
+)
+    return sum(-0.5 * (log(2π) .+ log.(lik.σ²) .+ ((y .- f_mean).^2 .+ f_var) / lik.σ²))
+end
+
+# The closed form solution for a Poisson likelihood with an exponential inverse link function
+function expected_loglik(
+    ::Analytic,
+    y::AbstractVector,
+    f_mean::AbstractVector,
+    f_var::AbstractVector,
+    ::PoissonLikelihood{ExpLink}
```

Comment:
not sure if we can find a better way of formulating it in code, but I did make a double-take and got a bit confused by the type parametrisation making it seem like it was an exponential link - not invlink - function.. what do the rest of you think ?

not helped by link functions being invertible ... I'm wondering if it'd be more helpful to somehow encode input/output domains (e.g. "ExpLink() isa LinkFunction{with input domain (-infty, +infty) and output domain (0, infty)"

Reply:
Well, I see it as an unfortunate legacy from statistics, it's just that it's very well defined that "f = link(expectation(y))" so when we think about transforming f to get the mean parameter e.g. for Poisson likelihood, then lambda = invlink(f). And `PoissonLikelihood{ExpLink}` then makes it seem like it might be a "log invlink" (not that that makes any sense)...

Reply:
Could we just call it `Exp` ? It would be even nicer if we could just dispatch on `exp` itself... :thinking:

Reply:
Should we move this to an issue in GPLikelihoods?